### PR TITLE
prov/efa: use gdrcopy

### PR DESCRIPTION
**Pull request by wzamazon**
This PR contains commits that make EFA to gdrcopy

### CODE REVIEW

File: prov/efa/src/efa_mr.c [hunk 1]
```diff
@@ -372,6 +399,8 @@ static int efa_mr_reg_impl(struct efa_mr *efa_mr, uint64_t flags, void *attr)
 	if (!efa_mr->ibv_mr) {
 		EFA_WARN(FI_LOG_MR, "Unable to register MR: %s\n",
 				fi_strerror(-errno));
+		if (cuda_is_gdrcopy_enabled())
```
Comment (wckzhang) [inline prov/efa/src/efa_mr.c:402]: Don't you also need to check that the handle is set?

File: prov/efa/src/rxr/rxr_pkt_type_base.c [hunk 2]
```diff
@@ -256,22 +256,46 @@ ssize_t rxr_pkt_copy_data_to_rx_entry(struct rxr_ep *ep,
...
+	 */
+	if (efa_ep_is_cuda_mr(desc) && !cuda_is_gdrcopy_enabled()) {
+		efa_ep = container_of(ep->rdm_ep, struct efa_ep, util_ep.ep_fid);
+		use_p2p = efa_ep_use_p2p(efa_ep, desc);
+		if (use_p2p < 0)
+			return use_p2p;
+
+		if (use_p2p == 0) {
+			FI_WARN(&rxr_prov, FI_LOG_CQ, "Neither p2p nor gdrcopy is available,"
+				"thus libfabric has not way to copy received data to Nvidia GPU");
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_pkt_type_base.c:289]: has no* way

File: prov/efa/src/rxr/rxr_msg.c [hunk 3]
```diff
@@ -74,15 +74,36 @@ ssize_t rxr_msg_post_cuda_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_ent
...
+	if (tx_entry->total_len == 0 || cuda_is_gdrcopy_enabled()) {
 		pkt_type = delivery_complete_requested ? RXR_DC_EAGER_MSGRTM_PKT : RXR_EAGER_MSGRTM_PKT;
-		return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry,
-					 pkt_type + tagged, 0, 0);
+		max_eager_data_size = rxr_pkt_req_max_data_size(rxr_ep,
+								tx_entry->addr,
+								pkt_type + tagged,
+								tx_entry->fi_flags, 0);
+
+		if (max_eager_data_size > rxr_env.efa_max_gdrcopy_msg_size)
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_msg.c:88]: why not just do max_eager_data_size = MIN(max_eager_data_size, rxr_env.efa_max_gdrcopy_msg_size);

File: prov/efa/src/rxr/rxr_msg.c [hunk 4]
```diff
@@ -74,15 +74,36 @@ ssize_t rxr_msg_post_cuda_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_ent
...
+			max_eager_data_size = rxr_env.efa_max_gdrcopy_msg_size;
+
+		if (tx_entry->total_len <= max_eager_data_size) {
+			return rxr_pkt_post_ctrl(rxr_ep, RXR_TX_ENTRY, tx_entry,
+						 pkt_type + tagged, 0, 0);
+		}
+
+		if (tx_entry->total_len <= rxr_env.efa_max_gdrcopy_msg_size) {
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_msg.c:96]: I feel like this is ignoring the case where tx_entry->total_len = 0;

**Review by wckzhang — CHANGES_REQUESTED** [review]

**Comment by wckzhang** [comment]
Where is gdrcopy actually being called? I see the handle being set up but do we call cuda_copy_to_dev somewhere?

File: prov/efa/src/efa_mr.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/efa_mr.c]: added an assertion here.

File: prov/efa/src/rxr/rxr_pkt_type_base.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: reworded in new revision.

File: prov/efa/src/rxr/rxr_msg.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_msg.c]: used `MIN` in revision.

File: prov/efa/src/rxr/rxr_msg.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_msg.c:95]: I am not sure. When `total_len = 0`, we used EAGER protocol in the previous if branch (around line 90).

**Comment by wzamazon** [comment]
> Where is gdrcopy actually being called? I see the handle being set up but do we call cuda_copy_from_dev somewhere? Is it the ofi_copy_from_hmem_iov as part of - rxr_pkt_init_data_from_tx_entry?

`ofi_copy_to_hmem_iov` will call `cuda_copy_to_dev`, which call `cuda_gdrcopy_to_dev`.

File: prov/efa/src/rxr/rxr_msg.c (hunk 4 above)
Comment (wckzhang) [reply prov/efa/src/rxr/rxr_msg.c:95]: Ah my bad, line 91 is a return, ignore

File: prov/efa/src/efa_mr.c [hunk 5]
```diff
@@ -372,6 +399,11 @@ static int efa_mr_reg_impl(struct efa_mr *efa_mr, uint64_t flags, void *attr)
 	if (!efa_mr->ibv_mr) {
 		EFA_WARN(FI_LOG_MR, "Unable to register MR: %s\n",
 				fi_strerror(-errno));
+		if (cuda_is_gdrcopy_enabled()) {
+			assert(efa_mr->peer.device.gdrcopy_handle);
```
Comment (wckzhang) [inline prov/efa/src/efa_mr.c:403]: I don't think this assertion is correct. cuda_is_gdrcopy_enabled can return true when attr->iface == FI_HMEM_SYSTEM, which means gdrcopy_handle wouldn't be set.

Comment (wzamazon) [reply prov/efa/src/efa_mr.c:403]: Thank you! added `efa_mr->peer.iface == FI_HMEM_CUDA` to the if condition.

**Review by wckzhang — APPROVED** [review]

**Comment by wzamazon** [comment]
Fixed two bugs in new revision:

1. use `cuda_dev_register` and `cuda_dev_unregister` directly, (instead of `cuda_gdrcopy_dev_register`) to avoid compilation issue.
2. set `long_cts` flag in tx_entry when long cts protocol is used (the flag is needed for delivery complete).

**Review by wckzhang — APPROVED** [review]

**Pull request merged**